src/DssProxyActions.sol

```diff
@@ -899,76 +819,72 @@ contract DssProxyActionsEnd is Common {
+        uint256 amt = _mul(wad, EndLike(end).fix(ilk)) / RAY / 10 ** (18 - GemJoinLike(gemJoin).dec());
         GemJoinLike(gemJoin).exit(msg.sender, amt);
     }
 }
 
 contract DssProxyActionsDsr is Common {
+    PotLike immutable public pot;
+
+    constructor(address daiJoin_, address pot_) public Common(daiJoin_) {
+        pot = PotLike(pot_);
+    }
+
     function join(
-        address daiJoin,
-        address pot,
-        uint wad
+        uint256 wad
     ) public {
-        VatLike vat = DaiJoinLike(daiJoin).vat();
         // Executes drip to get the chi rate updated to rho == now, otherwise join will fail
-        uint chi = PotLike(pot).drip();
+        uint256 chi = pot.drip();
         // Joins wad amount to the vat balance
-        daiJoin_join(daiJoin, address(this), wad);
+        daiJoin_join(address(this), wad);
         // Approves the pot to take out DAI from the proxy's balance in the vat
         if (vat.can(address(this), address(pot)) == 0) {
-            vat.hope(pot);
+            vat.hope(address(pot));
         }
         // Joins the pie value (equivalent to the DAI wad amount) in the pot
-        PotLike(pot).join(mul(wad, RAY) / chi);
+        pot.join(_mul(wad, RAY) / chi);
     }
 
     function exit(
-        address daiJoin,
-        address pot,
-        uint wad
+        uint256 wad
     ) public {
-        VatLike vat = DaiJoinLike(daiJoin).vat();
         // Executes drip to count the savings accumulated until this moment
-        uint chi = PotLike(pot).drip();
+        uint256 chi = pot.drip();
         // Calculates the pie value in the pot equivalent to the DAI wad amount
-        uint pie = mul(wad, RAY) / chi;
+        uint256 pie = _mul(wad, RAY) / chi;
         // Exits DAI from the pot
-        PotLike(pot).exit(pie);
+        pot.exit(pie);
         // Checks the actual balance of DAI in the vat after the pot exit
-        uint bal = DaiJoinLike(daiJoin).vat().dai(address(this));
+        uint256 bal = vat.dai(address(this));
         // Allows adapter to access to proxy's DAI balance in the vat
         if (vat.can(address(this), address(daiJoin)) == 0) {
-            vat.hope(daiJoin);
+            vat.hope(address(daiJoin));
         }
         // It is necessary to check if due rounding the exact wad amount can be exited by the adapter.
         // Otherwise it will do the maximum DAI balance in the vat
-        DaiJoinLike(daiJoin).exit(
+        daiJoin.exit(
             msg.sender,
-            bal >= mul(wad, RAY) ? wad : bal / RAY
+            bal >= _mul(wad, RAY) ? wad : bal / RAY
         );
     }
 
-    function exitAll(
-        address daiJoin,
-        address pot
-    ) public {
-        VatLike vat = DaiJoinLike(daiJoin).vat();
+    function exitAll() public {
         // Executes drip to count the savings accumulated until this moment
-        uint chi = PotLike(pot).drip();
+        uint256 chi = pot.drip();
```

Comment:
`chi` is used once but affects what `join` and `exit` does. So we need to call `drip` before, that's why we cache `chi`. In the case of `exit` because we want to accrue fees until the very end, in the case of `join` because is mandatory.

Reply:
In this case `pie` value is actually used twice.